### ISSUE TITLE
`finalize_target` is slow on large recipes with remote filesystem #987

### ISSUE THREAD

**Issue by TomAugspurger** [pangeo-forge/pangeo-forge-recipes#254]
`finalize_target`, specifically https://github.com/pangeo-forge/pangeo-forge-recipes/blob/4f8c0522c1395aaf15a820cfd637a1f1e76729b2/pangeo_forge_recipes/recipes/xarray_zarr.py#L558, can be slow.

We call `set(group)` to figure out what which variables are available. I didn't realize it, but apparently that can trigger a full listing of the bucket / storage container, which is slow with many objects. (Zarr has a fastpath if the object implements `listdirs` https://github.com/zarr-developers/zarr-python/blob/f0677c2e051cad672e781c6bf63b5edfe57aaca4/zarr/storage.py#L163-L165), but fsspec mappers don't; The Azure portal does it somehow, so it might be feasible)

I wonder if we can determine the set of variables based on some other mechanism. For now, I'm going to manually run `finalize_target` with hard-coded variables for the dataset I'm working on (gpm-imerg).

cc @sharkinsspatial, I think that this is one of the reasons `finalize_target` was taking so long for you.

edit: Oh I just discoved the `walk_blobs` method in azure.storage.blob. Perahps fsspec could use `walk` to implement `listdirs`. I'll investigate and open an fsspec issue if it seems promising.

**Comment by TomAugspurger** [comment]
From Ryan in https://github.com/pangeo-data/pangeo-cloud-federation/issues/987#issuecomment-989181682

> Thanks for reporting Tom.

> This seems like something that we should try to fix upstream, in fsspec and / or the fsspec + Zarr interface.

> In the meantime, if listing the group is really slow, we could at least only do it once. I think the current code iterates through the group keys twice.

Agreed on both counts.

**Comment by martindurant** [comment]
> Zarr has a fastpath if the object implements listdirs https://github.com/zarr-developers/zarr-python/blob/f0677c2e051cad672e781c6bf63b5edfe57aaca4/zarr/storage.py#L163-L165), but fsspec mappers don't

Pretty sure the FSSpecStore does! I'll check...

**Comment by martindurant** [comment]
https://github.com/zarr-developers/zarr-python/blob/master/zarr/storage.py#L1236

**Comment by TomAugspurger** [comment]
Thanks @martindurant, I was using `fsspec.mapping.FSMap` (https://github.com/fsspec/filesystem_spec/blob/1f3b6d81feb3927d368727012823292e1da7cd2d/fsspec/mapping.py#L7) as returned by `AzureBlobFileSystem.get_mapper()`.

Should we copy that implementation to fsspec.mapping.FSMap?

That said, the `ls` at https://github.com/zarr-developers/zarr-python/blob/f0677c2e051cad672e781c6bf63b5edfe57aaca4/zarr/storage.py#L1239-L1240 concerns me.

And I see now that `zarr.storage.ABSStore` implements the optimized version using walk: https://github.com/zarr-developers/zarr-python/blob/f0677c2e051cad672e781c6bf63b5edfe57aaca4/zarr/_storage/absstore.py#L161-L169

**Comment by martindurant** [comment]
There are various implementations of ls, but it ought to be the right call for this specific action of finding the subdirectories of a directory (but not listing the inner files). Also, it is cached on the instance object, where find() (which is what list(mapper) does) is likely not. If you did want to get all the files, find() generally has optimised implementations, but falls back to repeatedly calling ls and walking the tree (and so may be pretty inefficient).

It is unfortunate that the call for azure is "walk_blobs", where walk is a different file system method.

**Comment by TomAugspurger** [comment]
Thanks. Any thoughts on whether pangeo-forge should use `self.fs.get_mapper()` vs. `zarr.storage.FSMap`?

One potential issue with using `zarr.storage.FSMap`, is that `pangeo-forge` passes around fsspec instances. zarr's `FSMap` might need to grow a `.from_filesystem` method to be easily usable by pangeo-forge. Alternatively, pangeo-forge would need to capture the keyword arguments used to create the filesystem.

**Comment by martindurant** [comment]
FSStore is made for this, but I don't think that pangeo-forge needs any of its features versus a mapper *except* for directory listing. I suppose that was one of the motivations for consolidated metadata.

I agree that from_mapper is a reasonable ting for FSStore to have, but the arguments to fsspec.get_mapper are compatible with FSSpec anyway, and the mapper and/or store don't have any additional state, so you can pass them around just as easily. To call fs.get_mapper (or call FSMap directly) is almost exactly the same, just in two steps.

**Comment by rabernat** [comment]
My understanding has been that if I call `zarr.open(fs.get_mapper())`, the FSMap would be coerced to an FSStore. If this is not the case, we should fix it!

Having these two very similar but distinct objects is very confusing! 😕

**Comment by martindurant** [comment]
> FSMap would be coerced to an FSStore

I don't think so. We should therefore probably add both from_filesystem and from_mapper class methods to support all possible inputs but always end up with the same instance type.

**Comment by rabernat** [comment]
> We should therefore probably add both from_filesystem and from_mapper class methods to support all possible inputs but always end up with the same instance type.

I would def be :+1: on that.

**Comment by TomAugspurger** [comment]
Trying to understand next steps here:

I think having both `fsspec.mapping.FSMap` and `zarr.storage.FSStore` is confusing. They both implement the mutable mapping interface and so both can be passed to `zarr.open`, but one might be better than the other. We should remove that confusion by having just one implementation. IMO, that implementation should probably be in `fsspec`, but that doesn't really matter to me.

In the meantime, I think pangeo-forge-recipes should update [FSSpecTarget.get_mapper](https://github.com/pangeo-forge/pangeo-forge-recipes/blob/master/pangeo_forge_recipes/storage.py#L97-L99). I think, however, FSSpecTarget will need to be updated to include the `storage_options` used to create the fsspec filesystem instance (so that they can be passed through to `zarr.storage.FSStore` as `**storage_options`).

**Comment by martindurant** [comment]
Probably zarr constitutes the majority of use of FSMap. However, there is a push in zarr to create a storage base class that all other stores depends on, so we cannot achieve that within fsspec. In that model, FSMap would be wrapped in a DictStore (I think).

>  think, however, FSSpecTarget will need to be updated to include the storage_options used to create the fsspec filesystem instance 

Of have an `FSStore.from_filesystem(path)/from_fsmap` method, which is super easy to implement. It would be reasonable, given that zarr already checks if a URL is fsspec-compatible, to check whether the location argument is already an fsspec instance and do the right thing. Or xarray could do this (which also cares about checking for fsspec URLs).

**Comment by rabernat** [comment]
> I think having both `fsspec.mapping.FSMap` and `zarr.storage.FSStore` is confusing. They both implement the mutable mapping interface and so both can be passed to `zarr.open`, but one might be better than the other. We should remove that confusion by having just one implementation.

Hard :+1:.

> In the meantime, I think pangeo-forge-recipes should update [FSSpecTarget.get_mapper](https://github.com/pangeo-forge/pangeo-forge-recipes/blob/master/pangeo_forge_recipes/storage.py#L97-L99).

I am fine with this workaround while we await an upstream fix. It will be a bit awkward, but all of the parameters needed to [initialize the FSStore](https://zarr.readthedocs.io/en/stable/api/storage.html#zarr.storage.FSStore) should be available from the `target.fs` object.

One intermediate option might be to add a class method to [zarr.storage.FSStore](https://github.com/zarr-developers/zarr-python/blob/af0c36b2a612579db6d4ecc05388d9dcb2b3b1c2/zarr/storage.py#L1085) to allow you to initialize an FSStore from an existing fsspec.AbstractFileSystem + a path.

**Comment by rabernat** [comment]
> One intermediate option might be to add a class method to [zarr.storage.FSStore](https://github.com/zarr-developers/zarr-python/blob/af0c36b2a612579db6d4ecc05388d9dcb2b3b1c2/zarr/storage.py#L1085) to allow you to initialize an FSStore from an existing fsspec.AbstractFileSystem + a path.

I am starting a PR for this to zarr-python. I think it's the best path forward.

**Comment by rabernat** [comment]
With https://github.com/zarr-developers/zarr-python/pull/911 we should be able to add the following method to  `FSSpecTarget`

```python
    def get_fsstore(self) -> zarr.storage.FSStore:
        """Get a Zarr FSStore object suitable for storing Zarr data."""
        return zarr.storage.FSStore(self.root_path, fs=self.fs)
```

**Comment by alxmrs** [comment]
I suspect I just hit this issue, since my Dataflow job wasn't able to finish the `finalize_target()` step (traces are unclear). How can I help? 

> I am fine with this workaround while we await an upstream fix. It will be a bit awkward, but all of the parameters needed to initialize the FSStore should be available from the target.fs object.

Maybe we could push a simpler workaround in the meantime? How about updating the step this way, as proposed by @rabernat here (https://github.com/pangeo-forge/pangeo-forge-recipes/issues/214#issuecomment-930637189): 
```
for dim in dims: 
    if dim in group:
        arr = group[dim] 
```

As I understand it, the `__contains__` operation should avoid having to scan the whole filesystem we have with `set(group)`.  How does this sound? I'll experiment with this solution in a test branch right now.
